[CoordinatedGraphics] Use Damage to track the dirty region of GraphicsLayer

carlosgcampos opened the pull request:
#### c5f315a9682101aa5a3c1bb2f66dab405dacc7a4
<pre>
[CoordinatedGraphics] Use Damage to track the dirty region of GraphicsLayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=290385">https://bugs.webkit.org/show_bug.cgi?id=290385</a>

Reviewed by Nikolas Zimmermann.

We can use the Damage class to track the dirty rects in GraphicsLayer
which should improve the performance. This patch adds a viewSize()
method to GraphicsLayerClient since it&apos;s needed to create the Damage
from GraphicsLayer.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::mode const):
(WebCore::Damage::add):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::viewSize const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::setDirtyRegion):
(WebCore::CoordinatedPlatformLayer::setDamage): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::setNeedsDisplay):
(WebCore::GraphicsLayerCoordinated::setNeedsDisplayInRect):
(WebCore::GraphicsLayerCoordinated::commitLayerChanges):
(WebCore::GraphicsLayerCoordinated::updateDamage): Deleted.
(WebCore::GraphicsLayerCoordinated::updateDirtyRegion): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::viewSize const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::viewSize const):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Mode)):
(TestWebKitAPI::TEST(Damage, Move)):
(TestWebKitAPI::TEST(Damage, AddRect)):
(TestWebKitAPI::TEST(Damage, AddDamage)):
(TestWebKitAPI::TEST(Damage, Unite)):
(TestWebKitAPI::TEST(Damage, Resize)): Deleted.

Canonical link: <a href="https://commits.webkit.org/292747@main">https://commits.webkit.org/292747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0bfb1d58a43f972ef4edbe1e806a18a7ecce739

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102099 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47543 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16933 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25091 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73904 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31118 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100021 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12768 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/87748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54240 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12520 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5597 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46873 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5676 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104120 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24092 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17574 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82955 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24343 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83849 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82349 "Found 1 new API test failure: /TestWebKit:WebKit.OnDeviceChangeCrash (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27035 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/4579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17591 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15650 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24057 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29205 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23880 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27192 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25453 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -1092,8 +1057,11 @@ void GraphicsLayerCoordinated::commitLayerChanges(CommitState& commitState, floa
+    if (m_pendingChanges.contains(Change::DirtyRegion)) {
+        ASSERT(m_dirtyRegion.has_value());
+        m_platformLayer->setDirtyRegion(WTFMove(*m_dirtyRegion));
+        m_dirtyRegion = std::nullopt;
```
```suggestion
        m_platformLayer->setDirtyRegion(*std::exchange(m_dirtyRegion, std::nullopt));
```

I've checked this idiom, and it looks better than my first suggestion. Using std::exchange produces a prvalue T (the optional), then dereferencing it gets us an xvalue. 

I've done some grepping and this idiom is used already in the code: see with `git grep "*std::exchange" | grep std::nullopt`

carlosgcampos replied:
Ok, updated it in both places. Thanks!

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -75,6 +75,7 @@ class Damage {
     ALWAYS_INLINE bool isEmpty() const  { return m_rects.isEmpty(); }
+    ALWAYS_INLINE Mode mode() const { return m_mode; }
 
     void makeFull(const IntSize& size)
```
Side note: I dislike the name `makeFull` -- doesn't trigger any bells :-)
`extendToCoverWholeArea` would be more descriptive.

carlosgcampos replied:
There's a Full mode, so this makes the damage Full. Before there was a setMode but we don't really need to support changing from any mode to another, only make it full.

nikolaszimmermann commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -97,68 +98,62 @@ class Damage {
+            return true;
         }
 
         if (rectsCount == m_gridSize.unclampedArea()) {
```
In general comments would have helped here a bit to follow the logic -- you're just modifying existing code, I'm aware of that, just wanted to mention it.

nikolaszimmermann commented on Source/WebCore/platform/graphics/GraphicsLayerClient.h:
```diff
@@ -108,6 +108,8 @@ class GraphicsLayerClient {
     virtual float pageScaleFactor() const { return 1; }
     virtual float zoomedOutPageScaleFactor() const { return 0; }
 
+    virtual FloatSize viewSize() const { return { }; }
```
This confuses me -- what is a 'viewSize' of the GraphicsLayerClient? How does it differentiate from the "layer size" itself, etc.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -79,19 +79,27 @@ std::optional<PlatformLayerIdentifier> GraphicsLayerCoordinated::primaryLayerID(
+        if (m_dirtyRegion->mode() == Damage::Mode::Full)
+            return;
+
+        m_dirtyRegion->makeFull(m_size);
```
Would it make sense to move the `mode() == Damange::Mode::Full` check into the makeFull method?

carlosgcampos replied:
It's already there, but in this particular case I need to return early if it's already full.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -101,17 +109,16 @@ void GraphicsLayerCoordinated::setNeedsDisplayInRect(const FloatRect& initialRec
-    rects.append(rect);
-    noteLayerPropertyChanged(Change::DirtyRegion, ScheduleFlush::Yes);
+    if (!m_dirtyRegion) {
+        static constexpr FloatSize minDamageSize = { 512, 512 };
```
Why the 512 hardcoded here? Do you want to use the default tile size here?

Scony replied:
the bigger the damage rect, the bigger it's resolution i.e. it can store more damage rects thus describing the changed region in more precise way

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
```diff
@@ -101,17 +109,16 @@ void GraphicsLayerCoordinated::setNeedsDisplayInRect(const FloatRect& initialRec
+    if (!m_dirtyRegion) {
+        static constexpr FloatSize minDamageSize = { 512, 512 };
+        auto viewSize = client().viewSize();
+        m_dirtyRegion = Damage(m_size.constrainedBetween(minDamageSize, viewSize));
```
Doesn't that imply dirty regions are always at least 512x512 px wide? Basically always covering the whole tile (when default tile size is used?). That seems wrong.

carlosgcampos replied:
No, that's the minimum damage area we use, in this case it means that a grid of 2x2 is created, limiting the amount of dirty regions to 4 per tile.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -4503,6 +4503,17 @@ float RenderLayerCompositor::zoomedOutPageScaleFactor() const
     return page().zoomedOutPageScaleFactor();
 }
 
+FloatSize RenderLayerCompositor::viewSize() const
```
Ok, now it is clear what a viewSize() is -- how about enclosingFrameViewSize()?

carlosgcampos replied:
Or enclosingFrameViewVisibleSize()

nikolaszimmermann reviewed (changes requested):
Looks good.

carlosgcampos committed:
[CoordinatedGraphics] Use Damage to track the dirty region of GraphicsLayer
https://bugs.webkit.org/show_bug.cgi?id=290385

Reviewed by Nikolas Zimmermann.

We can use the Damage class to track the dirty rects in GraphicsLayer
which should improve the performance. This patch adds a viewSize()
method to GraphicsLayerClient since it's needed to create the Damage
from GraphicsLayer.

* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::mode const):
(WebCore::Damage::add):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::viewSize const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::setDirtyRegion):
(WebCore::CoordinatedPlatformLayer::setDamage): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::setNeedsDisplay):
(WebCore::GraphicsLayerCoordinated::setNeedsDisplayInRect):
(WebCore::GraphicsLayerCoordinated::commitLayerChanges):
(WebCore::GraphicsLayerCoordinated::updateDamage): Deleted.
(WebCore::GraphicsLayerCoordinated::updateDirtyRegion): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::viewSize const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::viewSize const):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Tools/TestWebKitAPI/Tests/WebCore/glib/Damage.cpp:
(TestWebKitAPI::TEST(Damage, Mode)):
(TestWebKitAPI::TEST(Damage, Move)):
(TestWebKitAPI::TEST(Damage, AddRect)):
(TestWebKitAPI::TEST(Damage, AddDamage)):
(TestWebKitAPI::TEST(Damage, Unite)):
(TestWebKitAPI::TEST(Damage, Resize)): Deleted.

Canonical link: https://commits.webkit.org/292747@main
